tools/contract-analyzer/analyzers/analyzers.go

```diff
@@ -0,0 +1,596 @@
+
+			inspector.Preorder(
+				elementFilter,
+				func(element ast.Element) {
+
+					type typeTuple struct {
+						first, second sema.Type
+					}
+					var typeTuples []typeTuple
+
+					switch element := element.(type) {
+					case *ast.ArrayExpression:
+						argumentTypes := elaboration.ArrayExpressionArgumentTypes[element]
+						if len(argumentTypes) < 2 {
+							return
+						}
+						typeTuples = append(
+							typeTuples,
+							typeTuple{
+								first:  argumentTypes[0],
+								second: argumentTypes[1],
+							},
+						)
+
+					case *ast.DictionaryExpression:
+						entryTypes := elaboration.DictionaryExpressionEntryTypes[element]
+						if len(entryTypes) < 2 {
+							return
+						}
+						typeTuples = append(
+							typeTuples,
+							typeTuple{
+								first:  entryTypes[0].KeyType,
+								second: entryTypes[1].KeyType,
+							},
+							typeTuple{
+								first:  entryTypes[0].ValueType,
+								second: entryTypes[1].ValueType,
+							},
+						)
+
+					case *ast.ConditionalExpression:
+						typeTuples = append(
+							typeTuples,
+							typeTuple{
+								first:  elaboration.ConditionalExpressionThenType[element],
+								second: elaboration.ConditionalExpressionElseType[element],
+							},
+						)
+
+					default:
+						return
+					}
+
+					for _, typeTuple := range typeTuples {
+						if typeTuple.first.Equal(typeTuple.second) {
+							continue
+						}
+
+						report(
+							analysis.Diagnostic{
+								Location: location,
+								Range:    ast.NewRangeFromPositioned(element),
+								Message:  "inferred type may differ",
+							},
+						)
+
+						// Only report one diagnostic for each expression
+						return
+					}
+
+				},
+			)
+
+			return nil
+		},
+	}
+})()
+
+func init() {
+	registerAnalyzer(
+		"supertype-inference",
+		SupertypeInferenceAnalyzer,
+	)
+}
+
+// External mutation analyzer
+
+var ExternalMutationAnalyzer = (func() *analysis.Analyzer {
+
+	elementFilter := []ast.Element{
+		(*ast.AssignmentStatement)(nil),
+		(*ast.MemberExpression)(nil),
+	}
+
+	outerCompositeTypes := func(
+		elaboration *sema.Elaboration,
+		stack []ast.Element,
+	) []*sema.CompositeType {
+
+		compositeTypes := make([]*sema.CompositeType, 0, len(stack))
+
+		for _, element := range stack {
+			compositeDeclaration, ok := element.(*ast.CompositeDeclaration)
+			if !ok {
+				continue
+			}
+
+			compositeType := elaboration.CompositeDeclarationTypes[compositeDeclaration]
+			compositeTypes = append(compositeTypes, compositeType)
+		}
+
+		return compositeTypes
+	}
+
+	memberExpressionMutatedElement := func(
+		elaboration *sema.Elaboration,
+		memberExpression *ast.MemberExpression,
+	) ast.Element {
+
+		memberInfo := elaboration.MemberExpressionMemberInfos[memberExpression]
+		member := memberInfo.Member
+		if member == nil {
+			return nil
+		}
+
+		switch memberInfo.AccessedType.(type) {
+		case *sema.DictionaryType:
+			switch member.Identifier.Identifier {
+			case "insert", "remove":
```

Comment:
The member resolvers themselves contain data about whether or not methods are considered "mutating": https://github.com/onflow/cadence/pull/1267/files#diff-c01765841e64b85306258f5ee4daff2bdc5b9e9e107167f55a6dba4162e8bdb8R1741

Is it possible to make use of this instead of hard-coding these?

Reply:
The analyzers are based on release v0.23, so this information is not available. Luckily there are only a few functions, and this is just a tool for developers, it is not used for execution.